src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java

```diff
@@ -998,4 +1032,293 @@ private static MethodHandle stringValueOf(Class<?> ptype) {
     private StringConcatFactory() {
         // no instantiation
     }
+
+    /**
+     * Simplified concatenation method to facilitate {@link StringTemplate}
+     * concatenation. This method returns a single concatenation method that
+     * interleaves fragments and values. fragment|value|fragment|value|...|value|fragment.
+     * The number of fragments must be one more that the number of ptypes.
+     * The total number of slots used by the ptypes must be less than or equal
+     * to MAX_INDY_CONCAT_ARG_SLOTS.
+     *
+     * @param fragments list of string fragments
+     * @param ptypes    list of expression types
+     *
+     * @return {@link MethodHandle}
+     *
+     * @throws StringConcatException If any of the linkage invariants are violated.
+     * @throws NullPointerException If any of the incoming arguments is null.
+     *
+     * @since 20
+     */
+    @PreviewFeature(feature=PreviewFeature.Feature.STRING_TEMPLATES)
+    public static MethodHandle makeConcatWithTemplate(
+            List<String> fragments,
+            List<Class<?>> ptypes)
+            throws StringConcatException
+    {
+        Objects.requireNonNull(fragments, "fragments is null");
```

Comment:
I think you need to do some defensive copy here
```
  ptypes = List.copyOf(pTypes);
```
to avoid the types and fragments to be changed at the same time they are checked.

Reply:
Changing